### ISSUE TITLE
Documentation Clarification

### ISSUE THREAD

**Issue by daraleigh** [fastly/terraform-provider-fastly#427]
The documentation states that the configuration id is optional:

Under Argument Reference:

```
configuration_id - (Optional) The ID of the set of TLS configuration options that apply to the enabled domains on this subscription.
```

When a request is submitted without the configuration ID an error comes back of 

```
Error: 400 - Bad Request:
Title: Bad request
Detail: Must provide TLS configuration relationship
```

If this is conditionally optional then it might serve the documentation to clarify what conditions make this field optional.

**Comment by Integralist** [comment]
We looked into this issue and unfortunately wasn't able to reproduce. 

In the Terraform provider acceptance tests we created a subscription without specifying the `configuration_id` and the tests pass without issue. 

I'm also not aware of any particular conditions in which it is mandatory. The API docs are also slightly confused as they say the field is "write-only on create", implying that the configuration can’t be queried after the subscription has been created, but then the response examples all show the field being returned. 

Maybe there's something funny going on with the behaviour being updated for some people and not others?

In any case, it would be good to understand whether you were doing anything special when you received the error. That might help to reproduce the issue and be able to have a closer look.

/cc @smaeda-ks

**Comment by smaeda-ks** [comment]
Closing for now as we don't have any further information to validate.

**Comment by rayterrill** [comment]
I'm running into this now unfortunately - any idea how to troubleshoot or work around this?

**Comment by Integralist** [comment]
Hi @rayterrill 

Are you able to produce a reduced test case of your Terraform configuration so I can better understand what you're trying to do with TLS Subscriptions.

Thanks!

**Comment by rayterrill** [comment]
Effectively this is what I have:
```
resource "fastly_service_vcl" "service" {
  count = var.enable_fastly ? 1 : 0

  name = myservice.mydomain.com

  dynamic "domain" {
    for_each = local.external_fqdns

    content {
      name = domain.value
    }
  }

  backend {
    address         = mybackend.mydomain.com
    name             = mybackend.mydomain.com
    port               = 443

    shield          = null

    #mutualtls client block - this is client auth from fastly -> appgw
    ssl_ca_cert     = var.fastly_appgw_mutualtls ? tls_self_signed_cert.cacert.cert_pem : null
    ssl_client_cert = var.fastly_appgw_mutualtls ? tls_locally_signed_cert.client.cert_pem : null
    ssl_client_key  = var.fastly_appgw_mutualtls ? tls_private_key.clientkey.private_key_pem : null

    #tls block - fastly -> appgw
    use_ssl           = true
    ssl_cert_hostname = myservice.mydomain.com
    ssl_check_cert    = true
  }

  dynamicsnippet {
    name     = "nocache"
    type     = "recv"
    priority = 110
  }

  force_destroy = true
}

resource "fastly_service_dynamic_snippet_content" "nocache" {
  for_each = {
    for d in fastly_service_vcl.service[0].dynamicsnippet : d.name => d if (d.name == "nocache" && var.enable_fastly)
  }
  
  service_id = fastly_service_vcl.service[0].id
  snippet_id = each.value.snippet_id

  manage_snippets = true

  content = var.fastly_disable_cache ? "return(pass);" : ""
}

resource "fastly_tls_subscription" "tlssub" {
  count = var.enable_fastly ? 1 : 0

  domains               = [for domain in fastly_service_vcl.service[0].domain : domain.name]
  certificate_authority = "lets-encrypt"
}
```

this results in:
```
│ Error: 400 - Bad Request:
│
│     Title:  Bad request
│     Detail: Must provide TLS configuration relationship
│
│   with fastly_tls_subscription.tlssub[0],
│   on fastly.tf line 54, in resource "fastly_tls_subscription" "tlssub":
│   54: resource "fastly_tls_subscription" "tlssub" {
```

**Comment by rayterrill** [comment]
i'll also engage fastly support on this.

**Comment by Integralist** [comment]
Thanks @rayterrill that would be good 👍🏻 

It looks like you're missing some required configuration to get TLS Subscriptions working.

Have you seen our documentation for configuring TLS Subscriptions yet?
https://registry.terraform.io/providers/fastly/fastly/latest/docs/resources/tls_subscription

The documentation includes multiple examples, the first is the 'basic' example that configures two subdomains (e.g. a.example.com, b.example.com).

In order for TLS Subscriptions to be setup via Terraform there is a specific set of resources that need to be configured.

The workflow configures a `fastly_tls_subscription` resource, then a `aws_route53_record` resource for handling the creation of the 'challenge' DNS records (e.g. `_acme-challenge.a.example.com` and `_acme-challenge.b.example.com`). 

> **NOTE:** If you don't use AWS Route53 for your DNS, then you'll need to find another third-party related provider resource that lets you configure DNS via Terraform. But for our examples we use AWS Route53.

We have to configure the `fastly_tls_subscription_validation` resource, as that helps block any other resource until the challenge DNS records have been validated by Fastly.

Once the validation has been successful, the `fastly_tls_configuration` data source is configured to filter the available results looking for an appropriate TLS configuration object (this looks to be what your current error is). If that filtering process is successful, then the subsequent `aws_route53_record` resources (for configuring the subdomains) will be executed using the returned TLS configuration data.

I'd recommend reading through the documentation and reaching out to support@fastly.com if you need more assistance.

**Comment by rayterrill** [comment]
Got it - yeah I was hoping to step through this a piece at a time (starting with the subscription) to better understand how the pieces are related vs drop that entire config in there, but it sounds like there are dependencies here that will prevent that from working.

**Comment by rayterrill** [comment]
@Integralist built this out, same error. we use azure not aws, but translated those pieces and it's the same error. Looking at it, that makes sense because nothing is telling the `fastly_tls_subscription` to wait for any other resource to be configured, so it never even got to the step where it would actually build out the dns side of things. The dns validation records also depend on the tls_subscription to be created - that's what I was thinking when I was trying to do this one step at a time.

Let me see if I can dump the web call that's being made in the terraform to trigger the 400.

**Comment by rayterrill** [comment]
The request is:
```
POST /tls/subscriptions HTTP/1.1
Host: api.fastly.com
```

Definitely not sending anything in the configuration section. Slightly sanitized below:
```
{
 "data": {
  "type": "tls_subscription",
  "attributes": {
   "certificate_authority": "lets-encrypt"
  },
  "relationships": {
   "tls_domain": {
    "data": [
     {
      "type": "tls_domain",
      "id": "myservice1.mydomain.com"
     },
     {
      "type": "tls_domain",
      "id": "myservice2.mydomain.com"
     },
     {
      "type": "tls_domain",
      "id": "myservice3.mydomain.com"
     },
     {
      "type": "tls_domain",
      "id": "myservice4.mydomain.com"
     },
     {
      "type": "tls_domain",
      "id": "myservice15mydomain.com"
     }
    ]
   }
  }
 },
 "included": [
  {
   "type": "tls_domain",
   "id": "myservice1.mydomain.com",
   "attributes": {
    "type": ""
   }
  },
  {
   "type": "tls_domain",
   "id": "myservice2.mydomain.com",
   "attributes": {
    "type": ""
   }
  },
  {
   "type": "tls_domain",
   "id": "myservice3.mydomain.com",
   "attributes": {
    "type": ""
   }
  },
  {
   "type": "tls_domain",
   "id": "myservice4.mydomain.com",
   "attributes": {
    "type": ""
   }
  },
  {
   "type": "tls_domain",
   "id": "myservice5.mydomain.com",
   "attributes": {
    "type": ""
   }
  }
 ]
}
```

**Comment by Integralist** [comment]
Hi @rayterrill 

Yup it might be best to move this discussion into an official Fastly support ticket (via support@fastly.com) so they can pick it up and review what you have (as obviously this is a public channel and makes sharing/discussing details tricky).

That said, I believe the Azure equivalent to `aws_route53_record` is probably [`azurerm_dns_cname_record`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/dns_cname_record). So I presume you'll need to do something like the following (see code comments for details)...

```terraform
resource "fastly_service_vcl" "example" {
    # your configuration
}

# IMPORTANT: There is a small typo in the docs that I've corrected here. 
# In the docs, the [for] expression tries to reference `fastly_service_vcl.testing-tls.domain`.
# But that should be `fastly_service_vcl.example.domain`.
resource "fastly_tls_subscription" "example" {
  domains               = [for domain in fastly_service_vcl.example.domain : domain.name]
  certificate_authority = "lets-encrypt"
}

# Copied straight from Azure docs.
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

# Copied straight from Azure docs.
resource "azurerm_dns_zone" "example" {
  name                = "mydomain.com"
  resource_group_name = azurerm_resource_group.example.name
}

resource "azurerm_dns_cname_record" "domain_validation" {
    # IMPORTANT: We need to wait for the 'Create TLS Subscription' API call to complete. 
    # https://developer.fastly.com/reference/api/tls/subs/#create-tls-sub
    depends_on = [fastly_tls_subscription.example]

    # Extract the relevant DNS record information
    for_each = {
      # The following `for` expression (due to the outer {}) will produce an object with key/value pairs.
      # The 'key' is the domain name we've configured (e.g. a.mydomain.com, b.mydomain.com)
      # The 'value' is a specific 'challenge' object whose record_name matches the domain (e.g. record_name is _acme-challenge.a.mydomain.com).
      for domain in fastly_tls_subscription.example.domains :
      domain => element([
        for obj in fastly_tls_subscription.example.managed_dns_challenges :
        obj if obj.record_name == "_acme-challenge.${domain}" # We use an `if` conditional to filter the list to a single element
      ], 0)                                                   # `element()` returns the first object in the list which should be the relevant 'challenge' object we need
    }

    name                = each.value.record_name
    zone_name           = azurerm_dns_zone.example.name
    resource_group_name = azurerm_resource_group.example.name
    ttl                 = 60
    record              = each.value.record_value
}

# This is a resource that other resources can depend on if they require the certificate to be issued.
# NOTE: Internally the resource keeps retrying `GetTLSSubscription` until no error is returned (or the configured timeout is reached).
# https://developer.fastly.com/reference/api/tls/subs/#get-tls-sub
resource "fastly_tls_subscription_validation" "example" {
    subscription_id = fastly_tls_subscription.example.id
    depends_on      = [azurerm_dns_cname_record.domain_validation] # Here we're waiting for the DNS records to be created by Azure.
}

# This data source lists all available TLS configuration data
# It uses the `default` attribute to narrow down the list to just one object.
# If the filtered list has a length that is not exactly one element, you'll see an error returned.
# That single TLS configuration element is then returned.
data "fastly_tls_configuration" "example" {
    default    = true
    depends_on = [fastly_tls_subscription_validation.example]
}

# Once validation is complete and we've retrieved the TLS configuration data (see the above data source), 
# we can then create multiple subdomain records using the relevant Azsure resource configuration. 
# See below for an example.
resource "azurerm_dns_cname_record" "example" {
    name                = "example.mydomain.com"
    zone_name           = azurerm_dns_zone.example.name
    resource_group_name = azurerm_resource_group.example.name
    ttl                 = 300
    record              = element([for record in data.fastly_tls_configuration.example.dns_records : record.record_value if record.record_type == "CNAME"], 0)
}
```

**Comment by rayterrill** [comment]
Sounds good - that's already happening. I assumed it would be a TXT record vs a CNAME on the validation side, but either way, this config doesn't make sense to me - maybe I'm just not understanding how the pieces work together.

The `azurerm_dns_cname_record` resource is dependent on the `fastly_tls_subscription` resource - but the terraform never gets that far because the `fastly_tls_subscription` fails to create with that 400. Is there another dependency on `fastly_tls_subscription` that should be added?

Trying to go at it from both sides since this was reported already as bug once and haven't got anywhere with the support side yet.

**Comment by Integralist** [comment]
Hi @rayterrill 

re:

> the terraform never gets that far because the fastly_tls_subscription fails to create with that 400.

Are you able to share the whole `400 Bad Request` request and response from the debug logs (e.g. `TF_LOG=debug`)? and _redact_ anything sensitive of course.

re:

> I haven't got anywhere with the support side yet.

Do you have a support ticket number or anything like that? or do you mean you've contacted support@fastly.com and not had any response.

Thanks!

**Comment by rayterrill** [comment]
@Integralist I do have a support ticket - I believe they're going to reach out to you as well on this.

Tried to post the request and response here: https://github.com/fastly/terraform-provider-fastly/issues/427#issuecomment-1474406389

**Comment by rayterrill** [comment]
@Integralist pretty sure we figured out what the issue is. Our fastly tenant doesn't have the "default" tls configuration in place, we only have custom tls configurations. And i guess those need to be entered by tls support - there's no public API for them, thus the confusion. I'll go ahead and see if I can use the existing tls configuration lookup data source to use one of the existing custom ones and plug that into the subscription to get this going.

**Comment by Integralist** [comment]
Ahh! OK great. Thanks for that update 👍🏻 

Hopefully one of the attributes from https://registry.terraform.io/providers/fastly/fastly/latest/docs/data-sources/tls_configuration#optional will help you to narrow down the list of configuration to the one you need to then create your DNS records.

**Comment by rayterrill** [comment]
@Integralist closing the loop here - got this working yesterday. indeed grabbing one of the tls configs and passing that into the tls sub resource worked great. potentially might be worth a doc update here to note that the tls subscription resource has an implicit dependency on the default/ootb tls configuration, and if you dont have that you'll need to handle that explicitly (i'm still not 100% on how this actually looks since we don't have that config - but hoping to save someone else time).

**Comment by Integralist** [comment]
Thanks @rayterrill much appreciated you taking the time to give feedback here 👍🏻

**Comment by rayterrill** [comment]
@Integralist wondering if there's a related issue here. I'm not able to update the tls subscription when it's provisioned, getting a 400:
```
 Error: 400 - Bad Request:
│
│     Title:  Bad request
│     Detail: Must provide TLS configuration relationship
│
│   with fastly_tls_subscription.tlssub[0],
│   on fastly.tf line 153, in resource "fastly_tls_subscription" "tlssub":
│  153: resource "fastly_tls_subscription" "tlssub" {
```

Looking at the go code here, it looks like there might again be an implicit dependency on the default tls configuration option?

And it looks like the existing code to build up the update doesn't account for the configuration id correct https://github.com/fastly/terraform-provider-fastly/blob/e0c80d39060082271f382965d21d341e1affaeb0/fastly/resource_fastly_tls_subscription.go#L328-L331 - it only adds that piece in if there's a change: https://github.com/fastly/terraform-provider-fastly/blob/e0c80d39060082271f382965d21d341e1affaeb0/fastly/resource_fastly_tls_subscription.go#L343-L345.

It looks like this is a PATCH so I'm surprised that's erroring out, but it does appear to be doing so. I can open a separate issue for this if you'd like.

**Comment by Integralist** [comment]
Hi @rayterrill yes please if you could open a fresh issue for this 👍🏻
